### ISSUE TITLE
Proper packaging for distributions

### ISSUE THREAD

**Issue by Geal** [sozu-proxy/sozu#161]
Right now, installing the proxy requires downloading rustup and installing the proxy through `cargo install sozu`, but we need a good packaging option for most Linux distributions.

As usual, whenever there's a new cool Rust toy available, @Keruspe [packages it for Exherbo](https://git.exherbo.org/summer/packages/net/sozu/index.html) before everybody else.

Which will be the next one?

- [ ] Ubuntu
- [ ] Debian
- [ ] Archlinux
- [ ] Redhat
- [ ] ...?

I'm especially interested in input about where to put elements like the default certificate, or the command unix socket.

**Comment by badboy** [comment]
I can take care of an AUR package for Arch Linux.

**Comment by Geal** [comment]
that would be great, thanks! Hopefully, something that could be automated at some point.

**Comment by badboy** [comment]
Would you like a package for a release or for the git version?

**Comment by Geal** [comment]
the release, I guess? I think I'll make a 0.3 soon, when I'm done breaking the API for now :)

**Comment by badboy** [comment]
Ok, now it just needs to compile

```
   Compiling sozu-lib v0.1.1
error: no associated item named `get_2048_256` found for type `openssl::dh::Dh` in the current scope
   --> /home/jer/.cargo/registry/src/github.com-1ecc6299db9ec823/sozu-lib-0.1.1/src/network/tls.rs:433:11
    |
433 |     match Dh::get_2048_256() {
    |           ^^^^^^^^^^^^^^^^

error: no method named `set_ecdh_auto` found for type `openssl::ssl::SslContextBuilder` in the current scope
   --> /home/jer/.cargo/registry/src/github.com-1ecc6299db9ec823/sozu-lib-0.1.1/src/network/tls.rs:441:13
    |
441 |     context.set_ecdh_auto(true);
    |             ^^^^^^^^^^^^^

error: no associated item named `get_2048_256` found for type `openssl::dh::Dh` in the current scope
   --> /home/jer/.cargo/registry/src/github.com-1ecc6299db9ec823/sozu-lib-0.1.1/src/network/tls.rs:587:11
    |
587 |     match Dh::get_2048_256() {
    |           ^^^^^^^^^^^^^^^^

error: no method named `set_ecdh_auto` found for type `openssl::ssl::SslContextBuilder` in the current scope
   --> /home/jer/.cargo/registry/src/github.com-1ecc6299db9ec823/sozu-lib-0.1.1/src/network/tls.rs:594:9
    |
594 |     ctx.set_ecdh_auto(true);
    |         ^^^^^^^^^^^^^

```

**Comment by badboy** [comment]
Seems to be a problem with having OpenSSL 1.1.0

**Comment by BlackYoup** [comment]
@badboy you can have a look at #178, sozu is not yet compatible with OpenSSL 1.1.0 indeed

**Comment by badboy** [comment]
Right, should have looked first. I [patched rust-openssl](https://github.com/badboy/rust-openssl/commit/49ceb328c760081bff55dc6c2e5ad61cb751169c) and now it compiles

**Comment by Geal** [comment]
oh, nice, thanks!

**Comment by Geal** [comment]
@badboy so, when can we move forward with this? :)

**Comment by badboy** [comment]
Sure! Initial PKGBUILD is here: https://github.com/badboy/aur-packages/blob/master/sozu-git/PKGBUILD

I'll add a default config, etc. today and will push it to the AUR then as well

**Comment by badboy** [comment]
Latest PKGBUILD here: https://github.com/badboy/aur-packages/blob/7a45fa0aef82ffaa7392cbc1fe0e50279f5b993f/sozu-git/PKGBUILD

Directory layout:

```
sozu-git /etc/
sozu-git /etc/sozu/
sozu-git /etc/sozu/config.toml
sozu-git /usr/
sozu-git /usr/bin/
sozu-git /usr/bin/sozu
sozu-git /usr/bin/sozuctl
sozu-git /usr/share/
sozu-git /usr/share/doc/
sozu-git /usr/share/doc/sozu-git/
sozu-git /usr/share/doc/sozu-git/README.md
sozu-git /usr/share/licenses/
sozu-git /usr/share/licenses/sozu-git/
sozu-git /usr/share/licenses/sozu-git/LICENSE-MIT
sozu-git /var/
sozu-git /var/lib/
sozu-git /var/lib/sozu/
```

State and the socket will be placed into `/var/lib/sozu`, though I think the socket should actually go into `/run/`

**Comment by msehnout** [comment]
Hi,
I'm working on a project of packaging Rust crates for Fedora (eventually for Red Hat Enterprise Linux):
https://docs.pagure.org/fedora-rust.sig/
The project is in very early stage, but we can also take a look at Sozu and try to package it.

**Comment by Geal** [comment]
@msehnout that would be great!

**Comment by Geal** [comment]
Ubuntu packages could be automated https://www.reddit.com/r/rust/comments/6o57em/distributing_binaries_through_ppas/

**Comment by Geal** [comment]
@msehnout did you take a look at https://github.com/sozu-proxy/sozu/pull/181 ? Now building RPM should work.

**Comment by Geal** [comment]
@badboy any chance you could send a PR for the arch PKGBUILD? Please put it in https://github.com/sozu-proxy/sozu/tree/master/os-build and add a script to build the package in one command

**Comment by badboy** [comment]
will take care of it tonight!

**Comment by msehnout** [comment]
@Geal I'll send a PR with few changes. I'm currently trying to make it build on my Fedora 26 machine, so I can build it on https://copr.fedorainfracloud.org/. The main problem now is that it downloads dependencies from crates.io during build time, which is fine for Copr repository, but all dependencies must be packaged into RPM if we want to get this package directly into Fedora repositories.

**Comment by Geal** [comment]
@msehnout take a look at cargo-vendor or cargo-local-registry for this. You can use it to download all dependencies in advance

**Comment by Geal** [comment]
debian is working on packaging rust crates: https://alioth-lists.debian.net/pipermail/pkg-rust-maintainers/2018-June/001412.html

**Comment by gjaegle** [comment]
any plan to deliver a docker container (under hub.docker.com)?

**Comment by NotBad4U** [comment]
@gjaegle  We already have one under [hub.docker.com](https://hub.docker.com/r/clevercloud/sozu/).
But we have some trouble with #470. I'll try to fix that as quickly as possible.

**Comment by clarfonthey** [comment]
FWIW, the AUR package is currently orphaned even though the repo seems to have an up-to-date version of the `PKGBUILD`. Someone from the project should probably adopt the AUR package and upload the version from the repo.

**Comment by FlorentinDUBOIS** [comment]
Closing — distribution packaging is in place: NixOS install docs (#1241), RPM spec fix (#1182), `os-build/` updates (#945). Reopen for any specific distro the current packaging doesn't cover.
